File: src/manuals.js
// Manual library: the PDFs live on the PC (the iPad cannot see its filesystem),
// so the bridge lists them and streams them back. Everything here is a pure
// function over a folder path, which is what makes it testable.

import fs from 'node:fs';
import path from 'node:path';

export const MAX_MANUALS = 600;
const PDF_EXTENSION = '.pdf';

/** Rejects anything that tries to leave the configured folder. */
export function resolveInside(root, relative) {
  if (typeof root !== 'string' || root.length === 0) return null;
  if (typeof relative !== 'string' || relative.length === 0) return null;
  // A NUL byte or an absolute/drive-qualified path is never a legitimate name.
  if (relative.includes('\0')) return null;
  if (path.isAbsolute(relative) || /^[a-zA-Z]:/.test(relative)) return null;
  const base = path.resolve(root);
  const target = path.resolve(base, relative);
  if (target !== base && !target.startsWith(`${base}${path.sep}`)) return null;
  return target;
}

/**
 * Walks the manual folder (sub-folders included) and returns the PDFs.
 * Unreadable sub-folders are skipped rather than failing the whole listing.
 */
export function listManuals(folder, { limit = MAX_MANUALS } = {}) {
  const result = { configured: false, folder: folder ?? '', files: [], truncated: false };
  if (typeof folder !== 'string' || folder.trim().length === 0) return result;
  result.folder = folder;
  let root;
  try {
    root = fs.realpathSync(folder);
  } catch {
    result.error = '找不到手册文件夹';
    return result;
  }
  result.configured = true;
  const found = [];
  const walk = (directory, depth) => {
    if (depth > 6 || found.length > limit) return;
    let entries;
    try {
      entries = fs.readdirSync(directory, { withFileTypes: true });
    } catch {
      return;
    }
    for (const entry of entries) {
      if (found.length > limit) return;
      if (entry.name.startsWith('.')) continue;
      const absolute = path.join(directory, entry.name);
      if (entry.isDirectory()) {
        walk(absolute, depth + 1);
        continue;
      }
      if (!entry.isFile() || path.extname(entry.name).toLowerCase() !== PDF_EXTENSION) continue;
      let stat;
      try {
        stat = fs.statSync(absolute);
      } catch {
        continue;
      }
      found.push({
        path: path.relative(root, absolute).split(path.sep).join('/'),
        name: path.basename(entry.name, PDF_EXTENSION),
        size: stat.size,
        modifiedAt: Math.round(stat.mtimeMs)
      });
    }
  };
  walk(root, 0);
  found.sort((a, b) => a.path.localeCompare(b.path, 'zh-Hans-CN'));
  result.truncated = found.length > limit;
  result.files = found.slice(0, limit);
  return result;
}

/** The absolute path of one manual, or null when it is not a servable PDF. */
export function manualFile(folder, relative) {
  const target = resolveInside(folder, relative);
  if (!target) return null;
  if (path.extname(target).toLowerCase() !== PDF_EXTENSION) return null;
  let stat;
  try {
    stat = fs.statSync(target);
  } catch {
    return null;
  }
  if (!stat.isFile()) return null;
  // Follow symlinks only inside the configured folder.
  try {
    const real = fs.realpathSync(target);
    const root = fs.realpathSync(folder);
    if (real !== root && !real.startsWith(`${root}${path.sep}`)) return null;
    return { absolute: real, size: stat.size };
  } catch {
    return null;
  }
}

/**
 * Single-range parser for the `Range: bytes=…` header. PDF.js asks for byte
 * ranges so it can jump into a large manual without downloading all of it.
 * Returns null when the header is absent or unusable (caller then sends 200).
 */
export function parseRange(header, size) {
  if (typeof header !== 'string' || !Number.isFinite(size) || size <= 0) return null;
  const match = /^bytes=(\d*)-(\d*)$/.exec(header.trim());
  if (!match) return null;
  const [, rawStart, rawEnd] = match;
  if (rawStart === '' && rawEnd === '') return null;
  let start;
  let end;
  if (rawStart === '') {
    // "bytes=-500" is the last 500 bytes.
    const length = Number(rawEnd);
    if (!Number.isFinite(length) || length <= 0) return null;
    start = Math.max(0, size - length);
    end = size - 1;
  } else {
    start = Number(rawStart);
    end = rawEnd === '' ? size - 1 : Number(rawEnd);
  }
  if (!Number.isInteger(start) || !Number.isInteger(end)) return null;
  if (start < 0 || start >= size) return null;
  if (end >= size) end = size - 1;
  if (end < start) return null;
  return { start, end, length: end - start + 1 };
}

File: windows-bridge/Manuals.cs
using System.Text.Json.Nodes;

namespace XPlaneEfbBridge;

/// <summary>
/// The PDF manual library.
///
/// The iPad cannot read this machine's disk, so the bridge lists the folder and
/// streams the files back. Everything here validates its input: a path that
/// escapes the configured folder, or a file that is not a PDF, is refused.
/// </summary>
internal static class Manuals
{
    public const int MaxManuals = 600;
    private const int MaxDepth = 6;

    /// <summary>Resolves a client-supplied relative path inside the folder, or null.</summary>
    public static string? ResolveInside(string folder, string relative)
    {
        if (string.IsNullOrWhiteSpace(folder) || string.IsNullOrWhiteSpace(relative)) return null;
        if (relative.Contains('\0')) return null;
        if (Path.IsPathRooted(relative) || relative.Length > 1 && relative[1] == ':') return null;
        string root;
        string target;
        try
        {
            root = Path.GetFullPath(folder);
            target = Path.GetFullPath(Path.Combine(root, relative));
        }
        catch
        {
            return null;
        }
        if (!target.StartsWith(root + Path.DirectorySeparatorChar, StringComparison.OrdinalIgnoreCase)
            && !string.Equals(target, root, StringComparison.OrdinalIgnoreCase)) return null;
        return target;
    }

    /// <summary>The absolute path of one manual, or null when it cannot be served.</summary>
    public static string? FilePath(string folder, string relative)
    {
        var target = ResolveInside(folder, relative);
        if (target is null) return null;
        if (!string.Equals(Path.GetExtension(target), ".pdf", StringComparison.OrdinalIgnoreCase)) return null;
        try
        {
            if (!File.Exists(target)) return null;
            // A symlink pointing outside the folder must not become a way out.
            var real = Path.GetFullPath(new FileInfo(target).LinkTarget is null ? target : target);
            var root = Path.GetFullPath(folder);
            return real.StartsWith(root, StringComparison.OrdinalIgnoreCase) ? real : null;
        }
        catch
        {
            return null;
        }
    }

    /// <summary>Lists the PDFs (sub-folders included) for the reader's file list.</summary>
    public static JsonObject List(string folder)
    {
        var payload = new JsonObject
        {
            ["configured"] = false,
            ["folder"] = folder ?? "",
            ["files"] = new JsonArray()
        };
        if (string.IsNullOrWhiteSpace(folder)) return payload;
        payload["folder"] = folder;
        if (!Directory.Exists(folder))
        {
            payload["error"] = "找不到手册文件夹";
            return payload;
        }
        payload["configured"] = true;
        var found = new List<(string Path, string Name, long Size, long Modified)>();
        void Walk(string directory, int depth)
        {
            if (depth > MaxDepth || found.Count > MaxManuals) return;
            IEnumerable<string> entries;
            try
            {
                entries = Directory.EnumerateFileSystemEntries(directory);
            }
            catch
            {
                return;                     // an unreadable sub-folder must not fail the listing
            }
            foreach (var entry in entries)
            {
                if (found.Count > MaxManuals) return;
                var name = Path.GetFileName(entry);
                if (name.StartsWith('.')) continue;
                try
                {
                    if (Directory.Exists(entry))
                    {
                        Walk(entry, depth + 1);
                        continue;
                    }
                    if (!string.Equals(Path.GetExtension(entry), ".pdf", StringComparison.OrdinalIgnoreCase)) continue;
                    var info = new FileInfo(entry);
                    if (!info.Exists) continue;
                    found.Add((Path.GetRelativePath(folder, entry).Replace('\\', '/'), Path.GetFileNameWithoutExtension(name), info.Length, new DateTimeOffset(info.LastWriteTimeUtc).ToUnixTimeMilliseconds()));
                }
                catch
                {
                    // Skip whatever we cannot stat.
                }
            }
        }
        Walk(folder, 0);
        found.Sort((a, b) => string.Compare(a.Path, b.Path, StringComparison.CurrentCulture));
        var files = new JsonArray();
        foreach (var item in found.Take(MaxManuals))
        {
            files.Add(new JsonObject
            {
                ["path"] = item.Path,
                ["name"] = item.Name,
                ["size"] = item.Size,
                ["modifiedAt"] = item.Modified
            });
        }
        payload["files"] = files;
        payload["truncated"] = found.Count > MaxManuals;
        return payload;
    }
}
